### PR TITLE
feat: Add opacity field to map tiles

### PR DESCRIPTION
**Pull request by KaWaite**
# Overview
closes reearth/reearth#277
Users want to be able to layer map tiles and change the opacity to see the tile below the top one.

## What I've done
Added a SliderField component to the right panel. It should be ok for other fields that need a slider as well, since its marks, min, max, step are all customizable with a "reasonable" fallback.

## What I haven't done
The SliderField steals lots of its logic from the Slider atom. There might be a way to consolidate into one component, but at the moment they are so different looking that it didn't seem worth the time to try. Keeping them as their own entities.

## How I tested
Added multiple map tiles and played around with different opacities. 
Change using slider.
Change using input (both with number pad and going up by steps with arrow keys on blur.

## Screenshot
Screen Shot 2022-05-09 at 17 08 06
Screen Shot 2022-05-09 at 17 08 35

### CODE REVIEW

File: src/components/molecules/EarthEditor/PropertyPane/PropertyField/SliderField/index.tsx [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+import { FieldProps } from "../types";
+
+export type Props = FieldProps<number> & {
+  min?: number;
+  max?: number;
+  step?: number;
+};
+
+const inputRegex = new RegExp(/^\d+\.\d{2,}$/);
```
Comment (rot1024) [inline src/components/molecules/EarthEditor/PropertyPane/PropertyField/SliderField/index.tsx:16]:
```suggestion
const inputRegex = /^\d+\.\d{2,}$/;
```

File: src/components/molecules/EarthEditor/PropertyPane/PropertyField/SliderField/index.tsx [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+  const isEditing = useRef(false);
+  const isDirty = useRef(false);
+
+  const calculatedStep = step ? step : max ? max / 10 : 0.1;
+
+  const opacityMarkers = {
+    [min ?? 0]: min ?? 0,
+    [max ? max / 2 : 0.5]: max ? max / 2 : 0.5,
+    [max ?? 1]: max ?? 1,
+  };
```
Comment (rot1024) [inline src/components/molecules/EarthEditor/PropertyPane/PropertyField/SliderField/index.tsx:39]:
```suggestion
  const opacityMarkers = useMemo(() => ({
    [min ?? 0]: min ?? 0,
    [max ? max / 2 : 0.5]: max ? max / 2 : 0.5,
    [max ?? 1]: max ?? 1,
  }), [min, max]);
```

useMemo is just helpful here

File: src/components/molecules/EarthEditor/PropertyPane/PropertyField/SliderField/index.tsx [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+          onChange(floatValue);
+          isDirty.current = false;
+        }
+      }
+    },
+    [onChange, min, max, value],
+  );
+
+  const handleInputChange = useCallback((e?: React.ChangeEvent<HTMLInputElement>) => {
+    const target = e?.currentTarget;
```
Comment (rot1024) [inline src/components/molecules/EarthEditor/PropertyPane/PropertyField/SliderField/index.tsx:73]:
```suggestion
  const handleInputChange = useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
    const target = e.currentTarget;
```

File: src/components/molecules/Visualizer/Engine/Cesium/hooks.ts [hunk 4]
```diff
@@ -51,24 +51,42 @@ export default ({
 
   // imagery layers
   const [imageryLayers, setImageryLayers] =
-    useState<[string, ImageryProvider, number | undefined, number | undefined][]>();
+    useState<
+      [string, ImageryProvider, number | undefined, number | undefined, number | undefined][]
```
Comment (rot1024) [inline src/components/molecules/Visualizer/Engine/Cesium/hooks.ts:55]: It's hard to understand the slice type when you get to this point. You can define a type:

```suggestion
      ImageryLayerData[]
```

```ts
type ImageryLayerData = {
  id: string;
  provider: ImageryProvider;
  min: number | undefined;
  max: number | undefined;
  opacity: number  | undefined;
}
```

**Review by rot1024 — CHANGES_REQUESTED** [review]

File: src/components/molecules/Visualizer/Engine/Cesium/hooks.ts [hunk 5]
```diff
@@ -51,24 +51,42 @@ export default ({
...
+        id,
+        type ? (url ? imagery[type](url) : imagery[type]()) : null,
+        min,
+        max,
+        opacity,
+      ])
       .filter(
-        (t): t is [string, ImageryProvider, number | undefined, number | undefined] => !!t[1],
+        (
+          t,
```
Comment (issmail-basel) [inline src/components/molecules/Visualizer/Engine/Cesium/hooks.ts:82]: @KaWaite  , can you explain this to me?

**Review by issmail-basel — APPROVED** [review]

**Review by rot1024 — APPROVED** [review]

**Pull request merged**